alf/networks/containers.py

```diff
@@ -0,0 +1,315 @@
+            name (str):
+        """
+        state_spec = [()] * len(layers)
+        for i, layer in enumerate(layers):
+            if isinstance(layer, Network):
+                state_spec[i] = layer.state_spec
+        if len(flatten(state_spec)) == 0:
+            state_spec = ()
+        self._state_spec = state_spec
+        if input_tensor_spec is None:
+            input_tensor_spec = get_input_tensor_spec(layers[0])
+        if input_tensor_spec is None:
+            raise ValueError(
+                "input_tensor_spec needs to "
+                "be provided for layers[0] of type %s" % type(layers[0]))
+        super().__init__(input_tensor_spec, name)
+        self._layers = nn.ModuleList(layers)
+
+    @property
+    def state_spec(self):
+        return self._state_spec
+
+    def forward(self, input, state=()):
+        x = input
+        if self._state_spec == ():
+            for layer in self._layers:
+                if isinstance(layer, Network):
+                    x = layer(x)[0]
+                else:
+                    x = layer(x)
+            return x, state
+        else:
+            new_state = [()] * len(self._layers)
+            for i, layer in enumerate(self._layers):
+                if isinstance(layer, Network):
+                    x, new_state[i] = layer(x, state[i])
+                else:
+                    x = layer(x)
+            return x, new_state
+
+    def copy(self, name=None):
+        """Create a copy of this network or return the current instance.
+
+        If ``self._singleton_instance`` is True, calling ``copy()`` will return
+        ``self``; otherwise it will make a copy of all the layers and re-initialize
+        their parameters.
+
+        Args:
+            name (str): name of the new network. Only used if not self._singleton_instance.
+        Returns:
+            Sequential:
+        """
+        if self._singleton_instance:
+            return self
+
+        if name is None:
+            name = self.name
+
+        new_layers = []
+        for l in self._layers:
+            if isinstance(l, Network):
+                layer = l.copy()
+            else:
+                layer = copy.deepcopy(l)
+                alf.layers.reset_parameters(layer)
+            new_layers.append(layer)
+        return Sequential(new_layers, self._input_tensor_spec, name)
+
+    def __getitem__(self, i):
+        return self._layers[i]
+
+
+class Parallel(Network):
+    """Apply each Network in the nest of Network to the corresponding input.
+
+    Example:
+
+    .. code-block:: python
+
+        net = Parallel((module1, module2))
+        y, new_state = net(x, state)
+
+    is equivalent to the following:
+
+    .. code-block:: python
+
+        y0, new_state0 = module1(x[0], state[0])
+        y1, new_state1 = module2(x[1], state[1])
+        y = (y0, y1)
+        new_state = (new_state0, new_state1)
+
+    """
+
+    def __init__(self, networks, input_tensor_spec=None, name="Parallel"):
```

Comment:
We should probably unify the argument names as either modules or networks. Right now Sequential call it layers.